File: project-java-cosminpac04/motorcycleContest/src/main/java/motorcycle/repository/TeamRepository.java
package motorcycle.repository;

import motorcycle.database.DataBaseConnection;
import motorcycle.model.Team;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;

public class TeamRepository implements IRepository<Team, Integer> {
    private static final Logger logger = LoggerFactory.getLogger(TeamRepository.class);

    @Override
    public void add(Team team) {
            String sql = "INSERT INTO teams (TeamName) VALUES (?)";
        try (Connection conn = (Connection) DataBaseConnection.connect();
             PreparedStatement stmt = conn.prepareStatement(sql)) {

            stmt.setString(1, team.getTeamName());
            stmt.executeUpdate();

            logger.info("Added team: {}", team.getTeamName());

        } catch (SQLException e) {
            logger.error("Error adding team: {}", team.getTeamName(), e);
        }
    }

    @Override
    public void delete(Team team) {
        String sql = "DELETE FROM teams WHERE Tid = ?";
        try (Connection conn = (Connection) DataBaseConnection.connect();
             PreparedStatement stmt = conn.prepareStatement(sql)) {

            stmt.setInt(1, team.getID());
            stmt.executeUpdate();

            logger.info("Deleted team with ID: {}", team.getID());

        } catch (SQLException e) {
            logger.error("Error deleting team with ID: {}", team.getID(), e);
        }
    }

    @Override
    public void update(Team team, Integer id) {
        String sql = "UPDATE teams SET TeamName = ? WHERE Tid = ?";
        try (Connection conn = (Connection) DataBaseConnection.connect();
             PreparedStatement stmt = conn.prepareStatement(sql)) {

            stmt.setString(1, team.getTeamName());
            stmt.setInt(2, id);
            stmt.executeUpdate();

            logger.info("Updated team with ID: {}", id);

        } catch (SQLException e) {
            logger.error("Error updating team with ID: {}", id, e);
        }
    }

    @Override
    public Team findById(Integer id) {
        String sql = "SELECT * FROM teams WHERE Tid = ?";
        try (Connection conn = (Connection) DataBaseConnection.connect();
             PreparedStatement stmt = conn.prepareStatement(sql)) {

            stmt.setInt(1, id);
            ResultSet rs = stmt.executeQuery();

            if (rs.next()) {
                Team team = new Team(rs.getInt("id"), rs.getString("team_name"));
                logger.info("Found team: {}", team.getTeamName());
                return team;
            } else {
                logger.warn("No team found with ID: {}", id);
            }

        } catch (SQLException e) {
            logger.error("Error finding team with ID: {}", id, e);
        }
        return null;
    }

    @Override
    public List<Team> findAll() {
        List<Team> teams = new ArrayList<>();
        String sql = "SELECT * FROM teams";

        try (Connection conn = (Connection) DataBaseConnection.connect();
             Statement stmt = conn.createStatement();
             ResultSet rs = stmt.executeQuery(sql)) {

            while (rs.next()) {
                teams.add(new Team(rs.getInt("id"), rs.getString("team_name")));
            }

            logger.info("Retrieved {} teams", teams.size());

        } catch (SQLException e) {
            logger.error("Error retrieving teams", e);
        }
        return teams;
    }

    @Override
    public List<Team> getAll() {
        return findAll();
    }
}

File: project-c-cosminpac04/motorcycleApp/Repositories/TeamRepository.cs
using System;
using System.Collections.Generic;
using System.Data.SQLite;
using motorcycleApp.DataBase;
using motorcycleApp.Models;
using Serilog;

namespace motorcycleApp.Repositories;

public class TeamRepository : IRepository<Team, int>
{
    public void Add(Team entity)
    {
        string sql = "INSERT INTO Team (TeamName) VALUES (@Name)";
        try
        {
            using (var conn = ConnectionManager.GetConnection())
            {
                Log.Information("Connection established: " + conn.Database);

                using (var command = new SQLiteCommand(sql, conn))
                {
                    command.Parameters.AddWithValue("@Name", entity.TeamName);
                    command.ExecuteNonQuery();
                    Log.Information("Added team '{TeamName}' to Database", entity.TeamName);
                }
            }
        }
        catch (Exception ex)
        {
            Log.Error(ex, "Error while adding team '{TeamName}' to Database", entity.TeamName);
        }
    }

    public void Update(Team entity)
    {
        string sql = "UPDATE Team SET TeamName=@Name WHERE ID=@Id";

        using (var conn = ConnectionManager.GetConnection())
        {
            Log.Information("Connection established: " + conn.Database);

            using (var command = new SQLiteCommand(sql, conn))
            {
                command.Parameters.AddWithValue("@Id", entity.ID);
                command.Parameters.AddWithValue("@Name", entity.TeamName);
                command.ExecuteNonQuery();
                Log.Information("Updated team ID {TeamId} to name '{TeamName}'", entity.ID, entity.TeamName);
            }
        }
    }

    public void Delete(Team entity)
    {
        string sql = "DELETE FROM Team WHERE ID=@Id";

        using (var conn = ConnectionManager.GetConnection())
        {
            Log.Information("Connection established: " + conn.Database);

            using (var command = new SQLiteCommand(sql, conn))
            {
                command.Parameters.AddWithValue("@Id", entity.ID);
                int affectedRows = command.ExecuteNonQuery();
                if (affectedRows > 0)
                {
                    Log.Information("Deleted team ID {TeamId} from Database", entity.ID);
                }
                else
                {
                    Log.Warning("No team found to delete with ID {TeamId}", entity.ID);
                }
            }
        }
    }

    public Team FindById(int id)
    {
        string sql = "SELECT * FROM Team WHERE ID=@Id";

        using (var conn = ConnectionManager.GetConnection())
        {
            Log.Information("Connection established: " + conn.Database);

            using (var command = new SQLiteCommand(sql, conn))
            {
                command.Parameters.AddWithValue("@Id", id);
                using (var reader = command.ExecuteReader())
                {
                    if (reader.Read())
                    {
                        var team = new Team(
                            reader.GetInt32(0),   // ID
                            reader.GetString(1)   // TeamName
                        );
                        Log.Information("Found team with ID {TeamId}: '{TeamName}'", team.ID, team.TeamName);
                        return team;
                    }
                }
            }
        }

        Log.Warning("Could not find team with ID {TeamId}", id);
        return null;
    }

    public IEnumerable<Team> GetAll()
    {
        string sql = "SELECT * FROM Team";
        var teams = new List<Team>();

        using (var conn = ConnectionManager.GetConnection())
        {
            Log.Information("Connection established: " + conn.Database);

            using (var command = new SQLiteCommand(sql, conn))
            using (var reader = command.ExecuteReader())
            {
                while (reader.Read())
                {
                    teams.Add(new Team(
                        reader.GetInt32(0),   // ID
                        reader.GetString(1)   // TeamName
                    ));
                }
            }
        }

        Log.Information("Retrieved {Count} teams from Database", teams.Count);
        return teams;
    }
}
